WIP support for parsing Git notes in the Pickable class when the --notes (-n) flag is provided

pratiksha-99 opened the pull request:
#### 6d9d1bd93a493c66ecb16689c68420f9c8f33353
<pre>
WIP support for parsing Git notes in the Pickable class when the --notes (-n) flag is provided


* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/local/git.py:
(Git):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/__init__.py:
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/annotate.py: Added.
(Annotate):
(Annotate.parser):
(Annotate.main):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pickable.py:
(Pickable.parser):
(Pickable.pickable):
(Pickable):
(Pickable.related_commits_from_notes):
(Pickable.main):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/pickable_unittest.py:
(TestPickable.test_double):
(TestPickable):
(TestPickable.test_related_commits_from_notes):
(TestPickable.test_related_commits_from_notes_none):
(TestPickable.test_related_commits_from_notes_no_notes):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/14c75ff025f7a3cb23548c241d3296f97a917ab3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102690 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22363 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12681 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107855 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53331 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/104729 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22672 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30865 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78100 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35073 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105696 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17564 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92676 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58432 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/102165 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17405 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10708 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52688 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87221 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10778 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110231 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29827 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21988 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87081 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-type-on-non-root.html (failure)") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/102245 "Passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30191 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88868 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86686 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31522 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9249 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/24080 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29754 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29562 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32889 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31124 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a25a54b7282424f3d999eeccab43f17950c3c3a8. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/44373

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/547da0d05468e3b4c831bb457809caf87763c780. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/44373

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/pickable_unittest.py:
```diff
@@ -317,3 +322,50 @@ def test_double(self):
+        commit.hash = 'commit_hash'
+
+        # Mock the `run` method used in the `related_commits_from_notes` method
+        related = Pickable.related_commits_from_notes(commit, run_command_function=mock_run)
```
This probably needs to be done with the much deeper mocks in `mocks/local/git.py`. I see you started to add some. These mocks are hard to setup, but they end up being expandable. @briannafan can hopefully help, 875467dc29efa0c3 is an example of a recently added mock

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pickable.py:
```diff
@@ -194,6 +215,32 @@ def pickable(cls, commits, repository, commits_story=None, excluded=None):
+        '''
+        Parse Git notes for patterns like:
+            Picked-from: abc1234
+            Cherry-picked-from: <hash>
```
Probably need something to exclude a commit too, right? Something like `Exclude-because:`

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pickable.py:
```diff
@@ -194,6 +215,32 @@ def pickable(cls, commits, repository, commits_story=None, excluded=None):
+
+        related = []
+        for line in stdout_str.strip().splitlines():
+            match = re.search(r'(Cherry-picked|Picked)-from:\s+([0-9a-f]{6,40})', line, re.IGNORECASE)
```
This is a pretty specific regex. First, it needs to support identifiers (there are a number of regexes that do both that you can copy, maybe even use directly). Second, because this is so specific, we probably need a `git-webkit skip` or `git-webkit annotate` that helps users craft such a note

pratiksha-99 committed:
WIP support for parsing Git notes in the Pickable class when the --notes (-n) flag is provided

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/local/git.py:
(Git):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/__init__.py:
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/annotate.py: Added.
(Annotate):
(Annotate.parser):
(Annotate.main):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pickable.py:
(Pickable.parser):
(Pickable.pickable):
(Pickable):
(Pickable.related_commits_from_notes):
(Pickable.main):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/pickable_unittest.py:
(TestPickable.test_double):
(TestPickable):
(TestPickable.test_related_commits_from_notes):
(TestPickable.test_related_commits_from_notes_none):
(TestPickable.test_related_commits_from_notes_no_notes):

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6d9d1bd93a493c66ecb16689c68420f9c8f33353. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/44373

pratiksha-99 commented:
I added the `annotate` support (still working on mocks for unit tests).
Example:
```
➜ git-webkit annotate 547da0d05468e3b4c831bb457809caf87763c780 --picked-from test-pqrstuvxyz
➜ git-webkit annotate 547da0d05468e3b4c831bb457809caf87763c780 --cherry-picked-from test-cherrypickcommit --append 

➜ git log 547da0d05468e3b4c831bb457809caf87763c780
commit 547da0d05468e3b4c831bb457809caf87763c780
Author: Pratiksha Choudhury <p_choudhury@apple.com>
Date:   Sun Apr 27 19:05:43 2025 -0700

    WIP support for printing Git notes in the Pickable class when the --notes (-n) flag is provided

Notes:
    Picked-from: test-pqrstuvxyz
    
    Cherry-picked-from: test-cherrypickcommit
```

pratiksha-99 committed:
print notes

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/14c75ff025f7a3cb23548c241d3296f97a917ab3. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/44373
